src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             }
-        }
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
+                    writeBytes >= totalBufferLength - activeBufferLength &&
+                    activeBufferLength > 0)
+                {
+                    // If the buffer has already grown to 32k, does not have room for the next request,
+                    // and is non-empty, flush the current contents to the wire.
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                _outgoingBuffer.EnsureAvailableSpace(writeBytes);
+            }
+            catch
+            {
+                if (acquireWriteLock)
+                {
+                    _writerLock.Release();
+                }
+                throw;
+            }
         }
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
+        // they want to release it.
+        private async ValueTask FinishWriteAsync(bool mustFlush, CancellationToken cancellationToken = default, bool releaseWriteLock = true)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
+                {
+                    Debug.Assert(_inProgressWrite == null);
+
+                    _inProgressWrite = FlushOutgoingBytesAsync();
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
```

Comment:
This is problematic in that until cancellationToken has cancellation requested (which may never happen), it's going to leak a registration from WaitWithCancellationAsync into the cancellation token.

At a minimum you should call tcs.TrySetResult(true), which will cause the WaitWithCancellationAsync method to complete and unregister from the token.

But this is also a fairly expensive call in general, especially since it's _inProgressWrite will already be completed.  The whole creation of the tcs and the if block should be surrounded by a check:
```C#
if (!_inProgressWrite.IsCompleted)
{
    ...
}
```
such that we only pay these costs if we actually have an asynchronously-completing write that hasn't yet completed.